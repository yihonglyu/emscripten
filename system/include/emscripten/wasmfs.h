/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#include <stdint.h>
#include <sys/stat.h>

#ifdef __cplusplus
extern "C" {
#endif

typedef struct Backend* backend_t;

// Obtains the backend_t of a specified path.
backend_t wasmfs_get_backend_by_path(char* path);

// Obtains the backend_t of a specified fd.
backend_t wasmfs_get_backend_by_fd(int fd);

// Creates a JSFile Backend in the new file system.
backend_t wasmfs_create_js_file_backend();

// Creates a Proxied Backend in the new file system.
backend_t wasmfs_create_proxied_backend(backend_t backend);

// Creates a new file in the new file system under a specific backend.
uint32_t wasmfs_create_file(char* pathname, mode_t mode, backend_t backend);

// Creates a new directory in the new file system under a specific backend.
long wasmfs_create_directory(char* path, long mode, backend_t backend);

#ifdef __cplusplus
}
#endif
