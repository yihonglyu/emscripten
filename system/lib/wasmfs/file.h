// Copyright 2021 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.

// This file defines the file object of the new file system.
// Current Status: Work in Progress.
// See https://github.com/emscripten-core/emscripten/issues/15041.

#pragma once

#include <assert.h>
#include <emscripten/html5.h>
#include <map>
#include <mutex>
#include <sys/stat.h>
#include <vector>
#include <wasi/api.h>

namespace wasmfs {

// Note: The general locking strategy for all Files is to only hold 1 lock at a
// time to prevent deadlock. This methodology can be seen in getDirs().

class Backend;
// This represents an opaque pointer to a Backend. A user may use this to
// specify a backend in file operations.
using backend_t = Backend*;
const backend_t NullBackend = nullptr;

class File : public std::enable_shared_from_this<File> {

public:
  enum FileKind { DataFileKind = 0, DirectoryKind, SymlinkKind };

  template<class T> bool is() const {
    static_assert(std::is_base_of<File, T>::value,
                  "File is not a base of destination type T");
    return int(kind) == int(T::expectedKind);
  }

  template<class T> std::shared_ptr<T> dynCast() {
    static_assert(std::is_base_of<File, T>::value,
                  "File is not a base of destination type T");
    if (int(kind) == int(T::expectedKind)) {
      return std::static_pointer_cast<T>(shared_from_this());
    } else {
      return nullptr;
    }
  }

  template<class T> std::shared_ptr<T> cast() {
    static_assert(std::is_base_of<File, T>::value,
                  "File is not a base of destination type T");
    assert(int(kind) == int(T::expectedKind));
    return std::static_pointer_cast<T>(shared_from_this());
  }

  ino_t getIno() {
    // Set inode number to the file pointer. This gives a unique inode number.
    // TODO: For security it would be better to use an indirect mapping.
    // Ensure that the pointer will not overflow an ino_t.
    static_assert(sizeof(this) <= sizeof(ino_t));
    return (ino_t)this;
  }

  backend_t getBackend() { return backend; }

  class Handle {

  protected:
    // This mutex is needed when one needs to access access a previously locked
    // file in the same thread. For example, rename will need to traverse
    // 2 paths and access the same locked directory twice.
    // TODO: During benchmarking, test recursive vs normal mutex performance.
    std::unique_lock<std::recursive_mutex> lock;
    std::shared_ptr<File> file;

  public:
    Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
    Handle(std::shared_ptr<File> file, std::defer_lock_t)
      : file(file), lock(file->mutex, std::defer_lock) {}
    bool trylock() { return lock.try_lock(); }
    size_t getSize() { return file->getSize(); }
    mode_t& mode() { return file->mode; }
    time_t& ctime() { return file->ctime; }
    time_t& mtime() { return file->mtime; }
    time_t& atime() { return file->atime; }

    // Note: parent.lock() creates a new shared_ptr to the same Directory
    // specified by the parent weak_ptr.
    std::shared_ptr<File> getParent() { return file->parent.lock(); }
    void setParent(std::shared_ptr<File> parent) { file->parent = parent; }

    std::shared_ptr<File> unlocked() { return file; }
  };

  Handle locked() { return Handle(shared_from_this()); }

  std::optional<Handle> maybeLocked() {
    auto handle = Handle(shared_from_this(), std::defer_lock);
    if (handle.trylock()) {
      return Handle(shared_from_this());
    } else {
      return {};
    }
  }

protected:
  File(FileKind kind, mode_t mode, backend_t backend)
    : kind(kind), mode(mode), backend(backend) {}
  // A mutex is needed for multiple accesses to the same file.
  std::recursive_mutex mutex;

  virtual size_t getSize() = 0;

  mode_t mode = 0; // User and group mode bits for access permission.

  time_t ctime = 0; // Time when the file node was last modified.
  time_t mtime = 0; // Time when the file content was last modified.
  time_t atime = 0; // Time when the content was last accessed.

  FileKind kind;

  // Reference to parent of current file node. This can be used to
  // traverse up the directory tree. A weak_ptr ensures that the ref
  // count is not incremented. This also ensures that there are no cyclic
  // dependencies where the parent and child have shared_ptrs that reference
  // each other. This prevents the case in which an uncollectable cycle occurs.
  std::weak_ptr<File> parent;

  // This specifies which backend a file is associated with.
  backend_t backend;
};

class DataFile : public File {

  virtual __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) = 0;
  virtual __wasi_errno_t
  write(const uint8_t* buf, size_t len, off_t offset) = 0;

public:
  static constexpr FileKind expectedKind = File::DataFileKind;
  DataFile(mode_t mode, backend_t backend)
    : File(File::DataFileKind, mode, backend) {}
  virtual ~DataFile() = default;

  class Handle : public File::Handle {

    std::shared_ptr<DataFile> getFile() { return file->cast<DataFile>(); }

  public:
    Handle(std::shared_ptr<File> dataFile) : File::Handle(dataFile) {}
    Handle(Handle&&) = default;

    __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) {
      return getFile()->read(buf, len, offset);
    }
    __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) {
      return getFile()->write(buf, len, offset);
    }
  };

  Handle locked() { return Handle(shared_from_this()); }
};

class Directory : public File {
protected:
  // TODO: maybe change to vector?
  std::map<std::string, std::shared_ptr<File>> entries;
  // 4096 bytes is the size of a block in ext4.
  // This value was also copied from the existing file system.
  size_t getSize() override { return 4096; }

public:
  static constexpr FileKind expectedKind = File::DirectoryKind;
  Directory(mode_t mode, backend_t backend)
    : File(File::DirectoryKind, mode, backend) {}

  struct Entry {
    std::string name;
    std::shared_ptr<File> file;
  };

  class Handle : public File::Handle {
    std::shared_ptr<Directory> getDir() { return file->cast<Directory>(); }

  public:
    Handle(std::shared_ptr<File> directory) : File::Handle(directory) {}
    Handle(std::shared_ptr<File> directory, std::defer_lock_t)
      : File::Handle(directory, std::defer_lock) {}

    std::shared_ptr<File> getEntry(std::string pathName);

    void setEntry(std::string pathName, std::shared_ptr<File> inserted) {
      // Hold the lock over both functions to cover the case in which two
      // directories attempt to add the file.
      auto lockedInserted = inserted->locked();
      getDir()->entries[pathName] = inserted;
      // Simultaneously, set the parent of the inserted node to be this Dir.
      // inserted must be locked because we have to go through Handle.
      // TODO: When rename is implemented, ensure that the source directory has
      // been removed as a parent.
      // https://github.com/emscripten-core/emscripten/pull/15410#discussion_r742171264
      assert(!lockedInserted.getParent());
      lockedInserted.setParent(file);
    }

    void unlinkEntry(std::string pathName) {
      // The file lock must be held for both operations. Removing the child file
      // from the parent's entries and removing the parent pointer from the
      // child should be atomic. The state should not be mutated in between.
      auto unlinked = getDir()->entries[pathName]->locked();
      unlinked.setParent({});
      getDir()->entries.erase(pathName);
    }

    // Used to obtain name of child File in the directory entries vector.
    std::string getName(std::shared_ptr<File> target) {
      for (const auto& [key, value] : getDir()->entries) {
        if (value == target) {
          return key;
        }
      }

      return "";
    }

    int getNumEntries() { return getDir()->entries.size(); }

    // Return a vector of the key-value pairs in entries.
    std::vector<Directory::Entry> getEntries() {
      std::vector<Directory::Entry> entries;
      for (const auto& [key, value] : getDir()->entries) {
        entries.push_back({key, value});
      }
      return entries;
    }

#ifdef WASMFS_DEBUG
    void printKeys() {
      for (auto keyPair : getDir()->entries) {
        emscripten_console_log(keyPair.first.c_str());
      }
    }
#endif
  };

  Handle locked() { return Handle(shared_from_this()); }

  std::optional<Handle> maybeLocked() {
    auto handle = Handle(shared_from_this(), std::defer_lock);
    if (handle.trylock()) {
      return Handle(shared_from_this());
    } else {
      return {};
    }
  }
};

struct ParsedPath {
  std::optional<Directory::Handle> parent;
  std::shared_ptr<File> child;
};

// TODO: Should this return a locked parent handle or a pointer to the parent?
// Given a pathname, this function will return a locked parent directory and a
// pointer to the specified file.
ParsedPath getParsedPath(std::vector<std::string> pathParts,
                         long& err,
                         std::shared_ptr<File> forbiddenAncestor = nullptr);

// Obtains parent directory of a given pathname.
// Will return a nullptr if the parent is not a directory.
// Will error if the forbiddenAncestor is encountered while processing.
// If the forbiddenAncestor is encountered, err will be set to EINVAL and
// nullptr will be returned.
std::shared_ptr<Directory>
getDir(std::vector<std::string>::iterator begin,
       std::vector<std::string>::iterator end,
       long& err,
       std::shared_ptr<File> forbiddenAncestor = nullptr);

// Return a vector of the '/'-delimited components of a path. The first
// element will be "/" iff the path is an absolute path.
std::vector<std::string> splitPath(char* pathname);

} // namespace wasmfs
