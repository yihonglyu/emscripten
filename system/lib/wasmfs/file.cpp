// Copyright 2021 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.
// This file defines the file object of the new file system.
// Current Status: Work in Progress.
// See https://github.com/emscripten-core/emscripten/issues/15041.

#include "file.h"
#include "wasmfs.h"
#include <emscripten/threading.h>

namespace wasmfs {
//
// Directory
//
std::shared_ptr<File> Directory::Handle::getEntry(std::string pathName) {
  auto it = getDir()->entries.find(pathName);
  if (it == getDir()->entries.end()) {
    return nullptr;
  } else {
    return it->second;
  }
}
//
// Path Parsing utilities
//
ParsedPath getParsedPath(std::vector<std::string> pathParts,
                         long& err,
                         std::shared_ptr<File> forbiddenAncestor) {
  std::shared_ptr<Directory> curr;
  auto begin = pathParts.begin();

  if (pathParts.empty()) {
    err = -ENOENT;
    return ParsedPath{{}, nullptr};
  }

  // Check if the first path element is '/', indicating an absolute path.
  if (pathParts[0] == "/") {
    curr = wasmFS.getRootDirectory();
    begin++;
    // If the pathname is the root directory, return the root as the child.
    if (pathParts.size() == 1) {
      return ParsedPath{curr->locked(), curr};
    }
  } else {
    curr = wasmFS.getCWD();
  }

  for (auto pathPart = begin; pathPart != pathParts.end() - 1; ++pathPart) {
    // Find the next entry in the current directory entry
#ifdef WASMFS_DEBUG
    directory->locked().printKeys();
#endif
    auto entry = curr->locked().getEntry(*pathPart);

    if (forbiddenAncestor) {
      if (entry == forbiddenAncestor) {
        err = -EINVAL;
        return ParsedPath{{}, nullptr};
      }
    }

    // Requested entry (file or directory)
    if (!entry) {
      err = -ENOENT;
      return ParsedPath{{}, nullptr};
    }

    curr = entry->dynCast<Directory>();

    // If file is nullptr, then the file was not a Directory.
    // TODO: Change this to accommodate symlinks
    if (!curr) {
      err = -ENOTDIR;
      return ParsedPath{{}, nullptr};
    }

#ifdef WASMFS_DEBUG
    emscripten_console_log(*pathPart->c_str());
#endif
  }

  // auto parent = curr->dynCast<Directory>();

  auto child = curr->locked().getEntry(*(pathParts.end() - 1));
  return ParsedPath{curr->locked(), child};
}

std::shared_ptr<Directory> getDir(std::vector<std::string>::iterator begin,
                                  std::vector<std::string>::iterator end,
                                  long& err,
                                  std::shared_ptr<File> forbiddenAncestor) {

  std::shared_ptr<File> curr;
  // Check if the first path element is '/', indicating an absolute path.
  if (*begin == "/") {
    curr = wasmFS.getRootDirectory();
    begin++;
  } else {
    curr = wasmFS.getCWD();
  }

  for (auto it = begin; it != end; ++it) {
    auto directory = curr->dynCast<Directory>();

    // If file is nullptr, then the file was not a Directory.
    // TODO: Change this to accommodate symlinks
    if (!directory) {
      err = -ENOTDIR;
      return nullptr;
    }

    // Find the next entry in the current directory entry
#ifdef WASMFS_DEBUG
    directory->locked().printKeys();
#endif
    curr = directory->locked().getEntry(*it);

    if (forbiddenAncestor) {
      if (curr == forbiddenAncestor) {
        err = -EINVAL;
        return nullptr;
      }
    }

    // Requested entry (file or directory)
    if (!curr) {
      err = -ENOENT;
      return nullptr;
    }

#ifdef WASMFS_DEBUG
    emscripten_console_log(it->c_str());
#endif
  }

  auto currDirectory = curr->dynCast<Directory>();

  if (!currDirectory) {
    err = -ENOTDIR;
    return nullptr;
  }

  return currDirectory;
}

// TODO: Check for trailing slash, i.e. /foo/bar.txt/
// Currently any trailing slash is ignored.
std::vector<std::string> splitPath(char* pathname) {
  std::vector<std::string> pathParts;
  char newPathName[strlen(pathname) + 1];
  strcpy(newPathName, pathname);

  // TODO: Other path parsing edge cases.
  char* current;
  // Handle absolute path.
  if (newPathName[0] == '/') {
    pathParts.push_back("/");
  }

  current = strtok(newPathName, "/");
  while (current != NULL) {
    pathParts.push_back(current);
    current = strtok(NULL, "/");
  }

  return pathParts;
}
} // namespace wasmfs
