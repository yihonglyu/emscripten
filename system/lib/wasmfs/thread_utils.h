/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

#include <assert.h>
#include <emscripten.h>
#include <emscripten/threading.h>
#include <pthread.h>

#include <functional>
#include <thread>
#include <utility>

namespace wasmfs {

// Helper function to call a callable object on the main thread in a fully
// asynchronous manner.
//
// Normal proxying to the main thread runs into the possible issue of the main
// thread pumping the event queue while it blocks. That is necessary for e.g.
// WebGL proxying where both sync and async events must be received and run in
// order. With this class, only async events are allowed, and they are run in
// an async manner on the main thread, that is, with nothing else on the stack
// while they run (even if this event blocks on something, which causes the
// main event loop to run events, if another event like this arrives, it would
// be queued to run later, avoiding a collision).

template<class F> void invoke_on_main_thread_async(F&& f) {
  using function_type = typename std::remove_reference<F>::type;

  auto run_on_main_thread = [](void* f1) {
    // Once on the main thread, run as an event from the JS event queue
    // directly, so nothing else is on the stack when we execute. This means
    // we are no longer ordered with respect to synchronous proxied calls,
    // but that is ok in this case as all we care about are async ones.
    emscripten_async_call(
      [](void* f2) {
        auto f = static_cast<function_type*>(f2);
        (*f)();
        delete f;
      },
      f1,
      0);
  };

  // Proxy the call to the main thread.
  emscripten_async_run_in_main_runtime_thread(
    EM_FUNC_SIG_VI,
    static_cast<void (*)(void*)>(run_on_main_thread),
    // If we were passed in something we need to copy, copy it; we end up with
    // a new allocation here on the heap that we will free later.
    new function_type(std::forward<F>(f)));
}

// Helper class for generic sync-to-async conversion. Creating an instance of
// this class will spin up a pthread. You can then call invoke() to run code
// on that pthread. The work done on the pthread receives a callback method
// which lets you indicate when it finished working. The call to invoke() is
// synchronous, while the work done on the other thread can be asynchronous,
// which allows bridging async JS APIs to sync C++ code.
//
// This can be useful if you are in a location where blocking is possible (like
// a thread, or when using PROXY_TO_PTHREAD), but you have code that is hard to
// refactor to be async, but that requires some async operation (like waiting
// for a JS event).
class sync_to_async {

  // Public API
  //==============================================================================
public:
  // Pass around the callback as a pointer to a std::function. Using a pointer
  // means that it can be sent easily to JS, as a void* parameter to a C API,
  // etc., and also means we do not need to worry about the lifetime of the
  // std::function in user code.
  using Callback = std::function<void()>*;

  //
  // Run some work on thread. This is a synchronous (blocking) call. The thread
  // where the work actually runs can do async work for us - all it needs to do
  // is call the given callback function when it is done.
  //
  // Note that you need to call the callback even if you are not async, as the
  // code here does not know if you are async or not. For example,
  //
  //  instance.invoke([](emscripten::sync_to_async::Callback resume) {
  //    std::cout << "Hello from sync C++ on the pthread\n";
  //    (*resume)();
  //  });
  //
  // In the async case, you would call resume() at some later time.
  //
  // It is safe to call this method from multiple threads, as it locks itself.
  // That is, you can create an instance of this and call it from multiple
  // threads freely.
  //
  void invoke(std::function<void(Callback)> newWork);

  //==============================================================================
  // End Public API

private:
  std::unique_ptr<std::thread> thread;
  std::mutex mutex;
  std::condition_variable condition;
  std::function<void(Callback)> work;
  bool readyToWork = false;
  bool finishedWork;
  bool quit = false;
  std::unique_ptr<std::function<void()>> resume;
  std::mutex invokeMutex;

  // The child will be asynchronous, and therefore we cannot rely on RAII to
  // unlock for us, we must do it manually.
  std::unique_lock<std::mutex> childLock;

  static void* threadMain(void* arg) {
    emscripten_async_call(threadIter, arg, 0);
    return 0;
  }

  static void threadIter(void* arg) {
    auto* parent = (sync_to_async*)arg;
    if (parent->quit) {
      pthread_exit(0);
    }
    // Wait until we get something to do.
    parent->childLock.lock();
    parent->condition.wait(parent->childLock,
                           [&]() { return parent->readyToWork; });
    auto work = parent->work;
    parent->readyToWork = false;
    // Allocate a resume function, and stash it on the parent.
    parent->resume = std::make_unique<std::function<void()>>([parent, arg]() {
      // We are called, so the work was finished. Notify the caller.
      parent->finishedWork = true;
      parent->childLock.unlock();
      parent->condition.notify_one();
      // Look for more work. Doing this asynchronously ensures that we continue
      // after the current call stack unwinds (avoiding constantly adding to the
      // stack, and also running any remaining code the caller had, like
      // destructors). TODO: add an option to do a synchronous call here in some
      // cases, which would avoid the time delay caused by a browser setTimeout.
      emscripten_async_call(threadIter, arg, 0);
    });
    // Run the work function the user gave us. Give it a pointer to the resume
    // function.
    work(parent->resume.get());
  }

public:
  sync_to_async() : childLock(mutex) {
    // The child lock is associated with the mutex, which takes the lock as we
    // connect them, and so we must free it here so that the child can use it.
    // Only the child will lock/unlock it from now on.
    childLock.unlock();

    // Create the thread after the lock is ready.
    thread = std::make_unique<std::thread>(threadMain, this);
  }

  ~sync_to_async() {
    // Wake up the child to tell it to quit.
    invoke([&](Callback func) {
      quit = true;
      (*func)();
    });

    thread->detach();
  }
};

void sync_to_async::invoke(std::function<void(Callback)> newWork) {
  // Use the invokeMutex to prevent more than one invoke being in flight at a
  // time, so that this is usable from multiple threads safely.
  std::lock_guard<std::mutex> invokeLock(invokeMutex);

  // Send the work over.
  {
    std::lock_guard<std::mutex> lock(mutex);
    work = newWork;
    finishedWork = false;
    readyToWork = true;
  }
  condition.notify_one();

  // Wait for it to be complete.
  std::unique_lock<std::mutex> lock(mutex);
  condition.wait(lock, [&]() { return finishedWork; });
}

} // namespace wasmfs
