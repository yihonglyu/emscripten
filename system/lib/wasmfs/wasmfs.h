// Copyright 2021 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.
// This file defines the global state of the new file system.
// Current Status: Work in Progress.
// See https://github.com/emscripten-core/emscripten/issues/15041.

#pragma once

#include "backend.h"
#include "file.h"
#include "file_table.h"
#include <assert.h>
#include <emscripten/html5.h>
#include <map>
#include <mutex>
#include <sys/stat.h>
#include <vector>
#include <wasi/api.h>

namespace wasmfs {

class WasmFS {

  std::vector<std::unique_ptr<Backend>> backendTable;
  FileTable fileTable;
  std::shared_ptr<Directory> rootDirectory;
  std::shared_ptr<Directory> cwd;
  std::mutex mutex;

  // Private method to initialize root directory once.
  // Initializes default directories including dev/stdin, dev/stdout,
  // dev/stderr. Refers to the same std streams in the open file table.
  std::shared_ptr<Directory> initRootDirectory();

  // Initialize files specified by --preload-file option.
  void preloadFiles();

public:
  // Files will be preloaded in this constructor.
  // This global constructor has init_priority 100. Please see wasmfs.cpp.
  // The current working directory is initialized to the root directory.
  WasmFS() : rootDirectory(initRootDirectory()), cwd(rootDirectory) {
    preloadFiles();
  }

  // This get method returns a locked file table.
  // There is only ever one FileTable in the system.
  FileTable::Handle getLockedFileTable() {
    return FileTable::Handle(fileTable);
  }

  // Returns root directory defined on WasmFS singleton.
  std::shared_ptr<Directory> getRootDirectory() { return rootDirectory; };

  // For getting and setting cwd, a lock must be acquired. There is a chance
  // that two threads could be mutating the cwd simultaneously.
  std::shared_ptr<Directory> getCWD() {
    const std::lock_guard<std::mutex> lock(mutex);
    return cwd;
  };

  void setCWD(std::shared_ptr<Directory> directory) {
    const std::lock_guard<std::mutex> lock(mutex);
    cwd = directory;
  };

  // Utility functions that can get and set a backend in the backendTable.
  backend_t addBackend(std::unique_ptr<Backend> backend) {
    const std::lock_guard<std::mutex> lock(mutex);
    backendTable.push_back(std::move(backend));
    return backendTable.back().get();
  }
};

// Global state instance.
extern WasmFS wasmFS;

} // namespace wasmfs
