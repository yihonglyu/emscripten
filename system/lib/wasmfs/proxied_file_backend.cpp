// Copyright 2021 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.

// This file defines the Proxied File and Proxied Backend of the new file
// system. Current Status: Work in Progress. See
// https://github.com/emscripten-core/emscripten/issues/15041.

#include "backend.h"
#include "file.h"
#include "thread_utils.h"
#include "wasmfs.h"

namespace wasmfs {

// This class represents a file that is resident on a background pthread. All
// file operations are proxied to this dedicated thread.
class ProxiedFile : public DataFile {

  sync_to_async& proxy;
  std::shared_ptr<DataFile> backendFile;

  // Read and write operations are forwarded to the file residing on the
  // dedicated thread.
  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
    __wasi_errno_t result;
    proxy.invoke([&](sync_to_async::Callback resume) {
      result = backendFile->locked().write(buf, len, offset);
      (*resume)();
    });
    return result;
  }

  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
    __wasi_errno_t result;
    proxy.invoke([&](sync_to_async::Callback resume) {
      result = backendFile->locked().read(buf, len, offset);
      (*resume)();
    });
    return result;
  }

  // Querying the size of the Proxied File returns the size of the underlying
  // file on the dedicated thread.
  size_t getSize() override {
    size_t result;
    proxy.invoke([&](sync_to_async::Callback resume) {
      result = backendFile->locked().getSize();
      (*resume)();
    });
    return result;
  }

public:
  // A file with the chosen destination backend is created on the Proxied File's
  // dedicated thread.
  ProxiedFile(mode_t mode,
              backend_t srcBackend,
              backend_t destBackend,
              sync_to_async& proxy)
    : DataFile(mode, srcBackend), proxy(proxy) {
    proxy.invoke([&](sync_to_async::Callback resume) {
      backendFile = destBackend->createFile(mode);
      (*resume)();
    });
  }

  // The destructor must proxy to the pthread and release the file stored on
  // that pthread.
  ~ProxiedFile() {
    proxy.invoke([&](sync_to_async::Callback resume) {
      backendFile = nullptr;
      (*resume)();
    });
  }
};
class ProxiedBackend : public Backend {
  backend_t backend;
  // The proxy member manages a dedicated thread that the Proxied Backend uses
  // to create files and directories under.
  sync_to_async proxy;

public:
  ProxiedBackend(backend_t backend) : backend(backend) {}

  std::shared_ptr<DataFile> createFile(mode_t mode) override {
    return std::make_shared<ProxiedFile>(mode, this, backend, proxy);
  }

  std::shared_ptr<Directory> createDirectory(mode_t mode) override {
    return std::make_shared<Directory>(mode, this);
  }
};

// Create a proxied backend by supplying another backend.
extern "C" backend_t wasmfs_create_proxied_file_backend(backend_t backend) {
  return wasmFS.addBackend(std::make_unique<ProxiedBackend>(backend));
}

} // namespace wasmfs
