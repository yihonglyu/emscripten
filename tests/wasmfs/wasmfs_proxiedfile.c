/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <emscripten/emscripten.h>
#include <emscripten/wasmfs.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

void write_and_read(const char* msg, int fd) {
  char buf[200] = {};
  errno = 0;
  write(fd, msg, strlen(msg));
  assert(errno == 0);
  errno = 0;
  lseek(fd, 0, SEEK_SET);
  assert(errno == 0);
  read(fd, buf, sizeof(buf));
  assert(errno == 0);
  printf("%s\n", buf);
}

int main() {

  // Create proxied proxied backend using composition.
  backend_t JSBackend = wasmfs_create_js_file_backend();
  backend_t proxiedBackend = wasmfs_create_proxied_file_backend(JSBackend);

  // Create a new proxied file under root.
  int fd = wasmfs_create_file("/testfile", 0777, proxiedBackend);

  // Try writing to and reading from the file.
  const char* msg = "Test with a new proxied file\n";
  write_and_read(msg, fd);

  // Verify that the size of the proxied file is the same as the written buffer.
  struct stat file;
  assert(fstat(fd, &file) != -1);
  assert(file.st_size == strlen(msg));

  // Try appending to the end of the proxied file.
  write_and_read(msg, fd);

  // Verify that the size of the proxied file has increased.
  assert(fstat(fd, &file) != -1);
  assert(file.st_size == strlen(msg) * 2);

  // Check that the proxied file has the correct backend type.
  assert(wasmfs_get_backend_by_fd(fd) == proxiedBackend);

  close(fd);

  assert(unlink("/testfile") != -1);

  // Try creating a new proxied backend directory under root.
  assert(wasmfs_create_directory("/test-dir", 0777, proxiedBackend) != -1);

  // Try to create a new proxied file under this new directory.
  int fd2 = open("/test-dir/proxiedfile", O_RDWR | O_CREAT, 0777);

  // Try writing to and reading from the file.
  const char* msg2 =
    "Test with a proxied file created under a proxied directory\n";

  write_and_read(msg2, fd2);

  // Check that the proxied file has the correct backend type.
  assert(wasmfs_get_backend_by_fd(fd2) == proxiedBackend);

  // Try creating an in-memory file under a proxied backend directory.
  backend_t memoryBackend = wasmfs_get_backend_by_path("/");

  int fd3 = wasmfs_create_file("/test-dir/inmemoryfile", 0777, memoryBackend);

  const char* msg3 =
    "Test with an in-memory file created under a proxied directory\n";
  write_and_read(msg3, fd3);

  return 0;
}
