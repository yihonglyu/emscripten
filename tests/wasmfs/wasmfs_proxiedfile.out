Test with a new proxied file

Test with a new proxied file
Test with a new proxied file

Test with a proxied file created under a proxied directory

Test with an in-memory file created under a proxied directory
